### PR TITLE
Issue 5140: Ensure ControllerResolverFactory ensure the scheduled task is interrupted during shutdown.

### LINKED ISSUE #5140: Does ControllerResolverFactory always properly shutdown?
**Problem description**
ControllerResolverFactory has a shutdown method which is not being invoked by ControllerImpl

**Problem location**
ControllerImpl.close()

**Suggestions for an improvement**
Call shutdown in close.

### PR DESCRIPTION
**Pull request by shrids**
Signed-off-by: Sandeep <sandeep.shridhar@emc.com>

**Change log description**  
- Ensure ControllerNameResolver#shutdown cancels the future the scheduled future with `mayInterruptIfRunning `flag as true.
- Wait for upto 20 seconds for channel termination to complete.
- Ensure watermarking thread pool is shutdown with clientFactory.close().

**Purpose of the change**  
Fixes #5140 

**What the code does**  
- Ensure ControllerNameResolver#shutdown cancels the future the scheduled future with `mayInterruptIfRunning `flag as true.
- Wait for upto 20 seconds for channel termination to complete.
- Ensure watermarking thread pool is shutdown with clientFactory.close().

**How to verify it**  
All the existing tests should pass.

### CODE REVIEW

File: client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java [hunk 1]
```diff
@@ -159,7 +159,9 @@ static WireCommand readCommand(InputStream in, IoBuffer buffer) throws IOExcepti
         }
         
         public void stop() {
-            stop.set(true);
+            if (stop.getAndSet(true)) {
```
Comment (ravisharda) [inline client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java:162]: Should the get and set separated out here? 
* This method returns if `get` `stop` returns true. 
* Set is called after the callback is complete, using compareAndSet(false, true). 

This might be especially useful if `callback.connectionDropped() can fail with an exception. 

On the other hand, if the failure of `connectionDropped()` call can be safely discounted, the existing code'd be OK.

File: client/src/main/java/io/pravega/client/connection/impl/ConnectionPoolImpl.java [hunk 2]
```diff
@@ -186,6 +186,7 @@ public void pruneUnusedConnections() {
     }
 
     @Override
+    @Synchronized
```
Comment (ravisharda) [inline client/src/main/java/io/pravega/client/connection/impl/ConnectionPoolImpl.java:189]: Why is this change required?

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java [hunk 3]
```diff
@@ -244,8 +244,10 @@ private Segment getSegmentForRevisionedClient(String scope, String streamName) {
 
     @Override
     public void close() {
-        controller.close();
+        // wait for 5 seconds before forcibly terminating the watermarkReader threads.
+        ExecutorServiceHelpers.shutdown(watermarkReaderThreads);
```
Comment (ravisharda) [inline client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java:248]: nit: There is no timeout duration specified, unlike what the comment suggests. This comment might be referring on the default value used in the `shutdown` method. If someone changes that value, this comment would go stale. It might be useful rather to say something like `Wait for default timeout duration before...`

**Review by ravisharda — CHANGES_REQUESTED** [review]

File: client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java (hunk 1 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java:162]: yeah, the idea is the to ensure that we do not stop and already stopped/closed connection.

File: client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java (hunk 3 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/stream/impl/ClientFactoryImpl.java:248]: updated it.

File: client/src/main/java/io/pravega/client/connection/impl/ConnectionPoolImpl.java (hunk 2 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/connection/impl/ConnectionPoolImpl.java:189]: Yeah, there can be a `getClientConnection()` invocation which would update the map after a connection creation while this close() method is invoked.

**Review by ravisharda — APPROVED** [review]

**Comment by shrids** [comment]
System test with this changeset completed successfully: https://asd-nautilus-jenkins.isus.emc.com/job/test-pravega-continuous-1/941/

File: client/src/main/java/io/pravega/client/control/impl/ControllerResolverFactory.java [hunk 4]
```diff
@@ -201,7 +201,7 @@ public void start(Listener listener) {
         public void shutdown() {
             shutdown = true;
             if (scheduledFuture != null) {
-                scheduledFuture.cancel(false);
+                scheduledFuture.cancel(true);
```
Comment (andreipaduroiu) [inline client/src/main/java/io/pravega/client/control/impl/ControllerResolverFactory.java:204]: FYI, this has no effect. It is not implemented in the JRE.

**Review by andreipaduroiu — APPROVED** [review]

File: client/src/main/java/io/pravega/client/control/impl/ControllerResolverFactory.java (hunk 4 above)
Comment (shrids) [reply client/src/main/java/io/pravega/client/control/impl/ControllerResolverFactory.java:204]: This is a ScheduledFuture on which cancel() is invoked.
` public ScheduledFuture<?> schedule(Runnable command, long delay, TimeUnit unit);`
As I understand your comment applies only on a CompletableFuture. let me know...

Comment (andreipaduroiu) [reply client/src/main/java/io/pravega/client/control/impl/ControllerResolverFactory.java:204]: Ah I missed that one. Nevermind then :)

**Review by tkaitchuck — COMMENTED** [review]
Looks good
